README.md

aos-wlan-ansible-role
=====================

This Ansible Network role provides a set of platform dependent configuration management modules specifically designed for the ArubaOS (AOS) Mobility Master and Standalone Controllers.

Requirements
------------

* Python 2.7 or 3.5+
* Ansible 2.8.1 or later  
* Minimum supported AOS firmware version 8.0

Installation
------------

Through Github, use the following command. Use option `-f` to overwrite current role version:

```
ansible-galaxy install git+https://github.com/aruba/aos-wlan-ansible-role.git
```

Through Galaxy:

```
ansible-galaxy install arubanetworks.aos_wlan_role
```

Inventory Variables
--------------

The variables that should be defined in your inventory for your AOS host are:

* `ansible_host`: IP address of controller in `A.B.C.D` format  
* `ansible_user`: Username for controller in `plaintext` format  
* `ansible_password`: Password for controller in `plaintext` format  
* `ansible_connection`: Must always be set to `httpapi`  
* `ansible_network_os`: Must always be set to `aos`
* `ansible_httpapi_port`: Must always be set to `4343`
* `ansible_httpapi_use_ssl`: Set `True` as AOS uses port 4343 for REST  
* `ansible_httpapi_validate_certs`: Set `True` or `False` depending on if Ansible should attempt to validate      certificates

### Sample Inventories:

Sample `inventory.yml`:

```yaml
all:
  hosts:
    controller:
      ansible_host: 10.1.1.1
      ansible_user: admin
      ansible_password: password
      ansible_connection: httpapi
      ansible_network_os: aos
      ansible_httpapi_port: 4343
      ansible_httpapi_validate_certs: True
      ansible_httpapi_use_ssl: True
```

Sample `inventory.ini`:

```ini
aos_1 ansible_host=10.1.1.1 ansible_user=admin ansible_password=password ansible_connection=httpapi ansible_network_os=aos ansible_httpapi_port=4343 ansible_httpapi_validate_certs=True ansible_httpapi_use_ssl=True
```

Example Playbook
----------------

If role installed through [Github](https://github.com/aruba/aos-wlan-ansible-role)
set role to `aos-wlan-ansible-role`:

```yaml
    ---
    -  hosts: all
       roles:
        - role: aos-wlan-ansible-role
       tasks:
         - name: Create a radius server
           aos_api_config:
             method: POST
             config_path: /md/SLR
             data:
              - rad_server:
                  - rad_server_name: test-dot1x
                    rad_host:
                      host: 1.1.1.1
```

If role installed through [Galaxy](https://galaxy.ansible.com/arubanetworks/aos8_role)
set role to `arubanetworks.aos_wlan_role`:

```yaml
    ---
    -  hosts: all
       roles:
        - role: arubanetworks.aos_wlan_role
       tasks:
         - name: Create a radius server
           aos_api_config:
             method: POST
             config_path: /md/SLR
             data:
              - rad_server:
                  - rad_server_name: test-dot1x
                    rad_host:
                      host: 1.1.1.1
```

You can also find pre-written playbooks for reference in the **sample_playbooks** directory on the GitHub repository. There are multiple playbooks for various use-cases/tasks typically performed on the Mobility Master, using different modules available with this role. You can choose an intended playbook and use it to build your own playbooks. 

Contribution
-------
At Aruba Networks we're dedicated to ensuring the quality of our products, if you find any
issues at all please open an issue on our [Github](https://github.com/aruba/aos-wlan-ansible-role) and we'll be sure to respond promptly!


License
-------

Apache 2.0

Author Information
------------------
Jay Pathak (jayp193)  
Karthikeyan Dhandapani (kdhandapani)  


RELEASE-NOTES.md
# 0.1.0

## Notable Changes
* This is the initial release for the AOS WLAN Ansible Role.


# 0.2.0

## Notable Changes
* Added new module for using show command on AOS8 - aos_show_command


library/aos_vlan.py

#!/usr/bin/python3
'''
Module for changing vlan configuration
'''

# -*- coding: utf-8 -*-
#
# (C) Copyright 2020 Hewlett Packard Enterprise Development LP.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied. See the License for the
# specific language governing permissions and limitations
# under the License.

from __future__ import (absolute_import, division, print_function)
__metaclass__ = type


ANSIBLE_METADATA = {'metadata_version': '1.0',
                    'status': ['preview'],
                    'supported_by': 'community'}

DOCUMENTATION = """
---
module: aos_vlan
version_added: 2.8.1
short_description: Configure VLAN IDs on ArubaOS products like Mobility Master
                   and Mobility Controllers using AOS APIs
description: Module for creating and deleting VLAN IDs and getting a list of
             configured VLAN IDs on the Mobility Master and Mobility Controllers
options:
    vlan_name:
        description:
            - Name given to a Named VLAN
              Example: User-VLAN
              Example: Mgmt-VLAN
        required: false
        type: str
    vlan_id:
        description:
            - VLAN ID/IDs or ranvge of VLANs to be configured.
              Example: 5
              Example: 5,10
              Example: 5,10,15-20
        required: false
        type: str
    config_path:
        description:
            - Path in configuration hierarchy to the node the API call is applied to
              On managed device this will be restricted to /md, while on a
              stand-alone controller, this will be restricted to /mm and /mm/mynode.
        required: false
        type: str
    action:
        description:
            - An action to get, create and delete VLANs or Named VLANs
        required: true
        choices:
            - get
            - create
            - delete
        type: str
    type:
        description:
            - Type of VLAN to be retrieved during GET operation.
        required: true
        choices:
            - all
            - named_vlan
        type: str
"""
EXAMPLES = """
#Usage Examples
    - name: Create a VLAN
      aos_vlan:
       action: create
       vlan_id: 5
       config_path: /md/Boston

    - name: Create multiple VLANs
      aos_vlan:
       action: create
       vlan_id: 5,10, 15-20
       config_path: /md/Boston

    - name: Create a named VLAN
      aos_vlan:
       action: create
       vlan_name: User-VLAN
       vlan_id: 5
       config_path: /md/Boston

    - name: Create multiple VLANs under a named VLAN
      aos_vlan:
       action: create
       vlan_name: User-VLANs
       vlan_id: 5,10-15
       config_path: /md/Boston

  *********************************************************************************
  NOTE: Playbook for creating Named VLANs with vlan_name parameter in the playbook
        will over-ride existing configuration if named VLAN with the same name
        exists.However, if done so, existing VLANs under the named VLAN will become
        unnamed.
  *********************************************************************************

    - name: Delete a named VLAN and VLANs under the named VLAN (2 tasks)
      aos_vlan:               #STEP-1 Delete name and ID mapping of the named VLAN
       action: delete
       vlan_name: User-VLAN
       config_path: /md/Boston
    - aos_vlan:               #STEP-2 Delete the VLANs configured under above named VLAN
       action: delete
       vlan_id: 5,10-15
       config_path: /md/Boston

    - name: Delete multiple VLANs
      aos_vlan:
       action: delete
       vlan_ids: 5,10
       config_path: /md/Boston

    - name: Delete multiple VLANs
      aos_vlan:
       action: delete
       vlan_ids: 15-20
       config_path: /md/Boston

    - name: Get a list of named vlans
      aos_vlan:
       action: get
       type: named_vlan
       config_path: /md/Boston

    - name: Get a list of all VLANs from current as well as parent hierarchy
      aos_vlan:
       action: get
       type: all
       config_path: /md/Boston


"""
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.aos_http import AosApi

def get_vlan_list(vlan_id):
    vlan_id = vlan_id.replace(" ", "")
    vlan_id_list = sum(((list(range(*[int(b) + c for c, b in enumerate(a.split('-'))]))
                         if '-' in a else [int(a)]) for a in vlan_id.split(',')), [])
    return vlan_id_list

def main():
    module = AnsibleModule(
        argument_spec=dict(
            vlan_name=dict(required=False, type='str'),
            vlan_id=dict(required=False, type='str'),
            config_path=dict(required=True, type='str'),
            action=dict(required=False, type='str', choices=['get', 'create', 'delete']),
            type=dict(required=False, type='str', choices=['all', 'named_vlan'], default='all')
        ))
    vlan_name = module.params.get('vlan_name')
    vlan_id = module.params.get('vlan_id')
    action = module.params.get('action')
    type_vlan = module.params.get('type')
    config_path = module.params.get('config_path')
    api = AosApi(module)
    config_url = "/v1/configuration/object/vlan_id?config_path=" + str(config_path)

    if action == "create":
        if vlan_name:
            config_url = "/v1/configuration/object?config_path=" + str(config_path)
            data = {"vlan_name": [{"_action": "modify", "name": vlan_name}],
                    "vlan_range": {"_action": "modify", "WORD": vlan_id.replace(" ", "")},
                    "vlan_name_id":[{"_action":"modify", "name": vlan_name,
                                     "vlan-ids": vlan_id.replace(" ", "")}]}
            result, changed = api.post(url=config_url, data=data)
            resp = result['resp']
        else:
            resp = []
            vlan_id_list = get_vlan_list(vlan_id)
            for vlan in vlan_id_list:
                data = {"id": vlan}
                result, changed = api.post(url=config_url, data=data)
                resp.append(result['resp'])

        module.exit_json(changed=changed, response=resp,
                               response_code=result['code'])

    elif action == "delete":
        if vlan_id and vlan_name:
            module.fail_json(change=False, response= result['resp'],
                             msg="To delete named VLAN, first delete a valid vlan_name."
                                 " Then use the vlan_id in a subsequent task if"
                                 " you wish to remove the VLAN ID associated to the"
                                 " named VLAN.")
        elif vlan_name and vlan_id is None:
            config_url = "/v1/configuration/object?config_path=" + str(config_path)
            data = {"vlan_name_id": [{"_action":"delete", "name": vlan_name}],
                    "vlan_name": [{"_action": "delete", "name": vlan_name}]}
            result, changed = api.post(url=config_url, data=data)
            resp = result['resp']
        else:
            vlan_id_list = get_vlan_list(vlan_id)
            resp = []
            for vlan in vlan_id_list:
                data = {"id": int(vlan), "_action": "delete"}
                result, changed = api.post(url=config_url, data=data)
                resp.append(result['resp'])
        module.exit_json(changed=changed, response=resp, response_code=result['code'])

    elif action == "get":
        if type_vlan == "named_vlan":
            config_url = "/v1/configuration/object/vlan_name_id?config_path="+str(config_path)
        result = api.get(url=config_url)
        resp = result['resp']
        if "_data" in resp.keys():
            module.exit_json(changed=False, response=resp, response_code=result['code'],
                             msg="Response shows the VLANs configured on the "
                                 "given config_path along with the ones inherited "
                                 "from the hierarchy above")
        else:
            module.fail_json(changed=False, response=result['resp'],
                             response_code=result['code'])

    else:
        module.fail_json(changed=False, response=result['resp'],
                         response_code=result['code'],
                         msg=str("Verify the playbook based on the playbook example"
                                 " in the module documentation."))


if __name__ == '__main__':
    main()


httpapi_plugins/aos.py
#!/usr/bin/python3
'''
httpapi Ansible plugin to connect with Aruba OS
running on the Mobility Master and Standalone controllers
'''

# -*- coding: utf-8 -*-
#
# (C) Copyright 2020 Hewlett Packard Enterprise Development LP.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied. See the License for the
# specific language governing permissions and limitations
# under the License.

import json
from ansible.module_utils._text import to_text
from ansible.module_utils.connection import ConnectionError
from ansible.module_utils.six.moves.urllib.error import HTTPError
from ansible.plugins.httpapi import HttpApiBase

DOCUMENTATION = """
---
author: Aruba Networks
httpapi: aos
short_description: Use REST to push configs to AOS8 devices
description:
  - This ArubaOS module provides REST interactions with ArubaOS Mobility Master
    and standalone controllers
version_added: 2.8.1
"""

class HttpApi(HttpApiBase):
    def __init__(self, *args, **kwargs):
        super(HttpApi, self).__init__(*args, **kwargs)

    def login(self, username, password):
        path = '/v1/api/login?username='+username+'&password='+password
        method = 'GET'
        self.send_request(data=None, path=path, method=method)

    def logout(self):
        path = '/v1/api/logout'
        data = None
        method = 'GET'
        self.send_request(data=data, path=path, method=method)

    def send_request(self, data, headers={}, **message_kwargs):
        # Ensure Connection
        if not self.connection._connected:
            self.connection._connect()
        if self.connection._auth:
            headers["Cookie"] = self.connection._auth["Cookie"]
            if 'logout' not in message_kwargs['path']:
                sess_tok = self.connection._auth["Cookie"].split("SESSION=")[1]
                message_kwargs['path'] = message_kwargs['path'] + "&UIDARUBA=" + sess_tok

        response, response_data = self.connection.send(data=data, headers=headers,
                                                       path=message_kwargs['path'],
                                                       method=message_kwargs['method'])
        try:
            response_data = json.loads(to_text(response_data.read()))
        except ValueError:
            response_data = response_data.read()
        except AttributeError as arr:
            raise Exception(str(response) + str(arr))
        return self.handle_response(response, response_data)

    def update_auth(self, response, response_text):
        """Return per-request auth token.
        The response should be a dictionary that can be plugged into the
        headers of a request. The default implementation uses cookie data.
        If no authentication data is found, return None
        """
        try:
            cookie = json.loads(to_text(response_text.getvalue()))
            if '_global_result' in cookie.keys():
                if 'UIDARUBA' in cookie['_global_result'].keys():
                    return {'Cookie': "SESSION=" + cookie['_global_result']['UIDARUBA']}
        except Exception as err:
            pass
        return None

    def handle_response(self, response, response_data):
        if isinstance(response, HTTPError):
            if response_data:
                error_text = "Error: " + str(response.code) + ", " + str(response_data)

                raise ConnectionError(error_text, code=response.code)
            raise ConnectionError(to_text(response), code=response.code)

        if response:
            return response_data, response.code

        return response_data, None


module_utils/aos_http.py
#!/usr/bin/python

from __future__ import (absolute_import, division, print_function)
__metaclass__ = type

import json
from ansible.module_utils.connection import Connection
from ansible.module_utils.six.moves.urllib.parse import urlencode

# -*- coding: utf-8 -*-
#
# (C) Copyright 2020 Hewlett Packard Enterprise Development LP.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied. See the License for the
# specific language governing permissions and limitations
# under the License.

class HttpHelper(object):
    def __init__(self, module):
        self._module = module
        self._connection_obj = None

    @property
    def _connection(self):
        if not self._connection_obj:
            self._connection_obj = Connection(self._module._socket_path)
        return self._connection_obj

    def http_request(self, url, method, data=None):
        return self._connection.send_request(data=data, method=method, path=url)

class AosApi(HttpHelper):
    def __init__(self, module):
        super(AosApi, self).__init__(module)
        self.api_version = "/v1"

    def get_query_params(self):
        query_params = {}
        params = dict(
            config_path=self._module.params.get('config_path'),
            filter=json.dumps(self._module.params.get('filter')),
            sort=self._module.params.get('sort'),
            count=self._module.params.get('count'),
            limit=self._module.params.get('limit'),
            offset=self._module.params.get('offset'),
            total=self._module.params.get('total')
            )
        query_params.update({key: value for key, value in params.items() if value is not None})
        return query_params

    def get_url(self, url, params=None):
        if params:
            return self.api_version + url + '?' +  urlencode(params)
        else:
            return self.api_version + url

    def format_data(self):
        data = None
        params_data = self._module.params.get('data')
        if len(params_data) == 1:
            # data
            data = params_data[0]
        else:
            # multipart data
            data = {'_list': params_data}
        return data

    def write_mem(self):
        config_path = self._module.params.get('config_path')
        url = "/configuration/object/write_memory"
        config_url = self.get_url(url, params={'config_path': config_path})
        result, changed = self.post(url=config_url)
        status = False
        if 'Error' not in result['resp'] and result['code'] == 200:
            status = True
        return status

    def validate_response(self, resp_data):
        res = True
        pending = 0
        status_str = []

        def status_check(resp_data, res):
            # Recursively check status in response data
            # Returns True if all status check passes
            # else return False
            if not res:
                return False
            if isinstance(resp_data, list):
                for ele in resp_data:
                    res = status_check(ele, res)
            elif isinstance(resp_data, dict):
                if "_result" in resp_data.keys():
                    if resp_data["_result"]["status"] != 0:
                        status_str.append(resp_data["_result"]["status_str"])
                        res = False
                for ele, val in resp_data.items():
                    if isinstance(val, (dict, list)):
                        res = status_check(val, res)
            return res

        if "_global_result" in resp_data:
            if resp_data["_global_result"]["status"] == 0:
                res = status_check(resp_data, True)
                if "_pending" in resp_data["_global_result"]:
                    pending = resp_data["_global_result"]["_pending"]
            else:
                res = False

                if "_pending" in resp_data["_global_result"]:
                    pending = resp_data["_global_result"]["_pending"]
                status_str.append(resp_data["_global_result"]["status_str"])
        elif "Error" in resp_data:
            res = False
            status_str.append(resp_data["Error"])

        if status_str:
            status_str = ", ".join(status_str)
        else:
            status_str = ""

        return res, pending, status_str

    def get(self, url, data=None):
        res, code = self.http_request(url=url, method="GET", data=data)
        result = {'resp': res, 'code': code}
        return result

    def post(self, url, data={}):
        data = json.dumps(data)
        config_path = self._module.params.get('config_path')
        idempotency_url = self.get_url('/configuration/object/config',
                                      params={'config_path': config_path})

        # Idemptency logic contributed by author sachaboudjema
        # Description: Check configuration in a node path
        #              before and after config push
        # Date Referenced: May 01 2020
        before = self.get(idempotency_url)['resp']

        res, code = self.http_request(url=url, method="POST", data=data)
        result = {'resp': res, 'code': code}

        after = self.get(idempotency_url)['resp']

        changed = (before != after)

        return result, changed


library/aos_api_config.py

#!/usr/bin/python3
'''
#Aruba OS 8(AOS8) API module
'''

# -*- coding: utf-8 -*-
#
# (C) Copyright 2020 Hewlett Packard Enterprise Development LP.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied. See the License for the
# specific language governing permissions and limitations
# under the License.

from __future__ import (absolute_import, division, print_function)
__metaclass__ = type


ANSIBLE_METADATA = {'metadata_version': '1.0',
                    'status': ['preview'],
                    'supported_by': 'community'}

DOCUMENTATION = """
---
module: aos_api_config
version_added: 2.8.1
short_descriptions: REST API module for ArubaOS 8.X
description: This module provides a configuration mechanism of ArubaOS products like Mobility Master and
                   Mobility Controllers using AOS 8 API
options:
    api_object:
        description:
            - Name of the API endpoint for which you would like to fetch data
              from the system. List of API endpoints can  be obtained at:
              https://<mm-ip-address>:4343/api (use your Mobility Master's IP)
        required: true
        type: str

    method:
        description:
            - HTTP Method to be used for the API call
        required: true
        choices:
            - GET
            - POST
        type: str

    config_path:
        description:
            - Path in the hierarchy where the API call should be applied
        required: false
        type: str

    data:
        description:
            - list of dictionaries where each element of list is a key value pairs
              of api object and corresponding configuration
        required: false
        type: list

    commit:
        description:
            - If set to True, it does a write_memory to flash
        required: false
        type: bool

    filter:
        description:
            - Filter applied on the data received by a GET request
              There are two types of filters:
              1) Object Filter:
                 An object filter limits which objects or which sub_objects should
                 be present in the response. Only one filter can be applied per request.
                 Object filter syntax is as follows:
                 [{"OBJECT" : { "<oper>" : [ <parameters> ] } } ]

                 <oper>
                   - $eq (matches one of the values)
                   - $neq (does not match any of the values)

                   <parameters> Comma separated values which need to be
                                filtered based on the operation

              2) Data Filter:
                 It filters out the configuration elements configured on the system.
                 Any filter which is not an object filter is a data filter.
                 Data filter should be applied as follows:
                 [{"<param-name" : { "<oper>" : <list-of-values> } } ]

                 <param_name> Fully qualified name of the parameter on values of
                               which filter needs to be applied

                  <oper>
                    - $eq (matches one of the values)
                    - $neq (does not match any of the values)
                    - $gt (matches a value which is greater than the filter)
                    - $gte (matches a value which is greater than or equal to the filter)
                    - $lt (matches a value which is less than the filter)
                    - $lte (matches a value which is less than or equal to the filter)
                    - $nin (pattern does not match the filter. Opposite of $in)
                    - $in (pattern matches the filter value. E.g., if filter says "ap",
                           "default-ap" and "ap-grp1" will both match)
        required: false
        type: list

    sort:
        description:
            - Data from the GET response can be sorted based on a single field
              Sort syntax is as follws:
              <oper><key>

               <oper> Order in which you want the values
                 - "+" (for ascending)(*default)
                 - "-" (for descending)

                <key> Key is the parameter name on which sort must be applied.
                      It is in the form:
                      <objectname>.<param_name>
                      or
                      <objectname>.<subobject_ name>.<param_name>

                ( Example: +int_vlan.id )
                ( Example: +int_vlan.int_vlan_mtu.value )
        required: false
        type: str

    count:
        description:
            - The count modifier just returns the total count of an object for
              multi-instance object or for multi-instance sub- object rather than
              the actual details of the objects. This is particularly useful when
              you want to get only the number of instances in an object rather than
              the object details.
              Usage example: count: int_vlan.int_vlan_ip_helper
        required: false
        type: str

    limit:
        description:
            - The maximum number if instances of an object that should be put in a
              single reques
        required: false
        type: int

    offset:
        description:
            - This conveys the number of the entries from which the response should
              start the next data set. For example, offset value of 21 means that
              out of all the instances of an object, give me data from 21st object.
              Currently, we support offset value to be multiples of the limit field
              below plus one. For example, if limit is 20, then offset can take values
              of 1, 21, 41, 61, 81 etc.
        required: false
        type: int

    total:
        description:
            - The total or maximum number of values that a response should contain
        required: false
        type: int

"""
EXAMPLES = """
#Usage Examples
    - name: Create a radius server
      aos_api_config:
        method: POST
        config_path: /md/Boston
        data:
         - rad_server: #POST request where a single endpoint is configured
             - rad_server_name: test-dot1x
               rad_host:
                 host: 1.1.1.1

       commit: True
        data: { "rad_server_name": "test_dot1x" ,"rad_host": {"host": "1.1.1.1"} }

    - name: Add a new VLAN, use it as an Interface VLAN and commit the pending changes
      aos_api_config:
        method: POST
        config_path: /md/SLR
        data:   #Multipart POST request where multiple endpoints are configured at once
         - vlan_id:
             - id: 45
         - int_vlan:
             - id: 45
               int_vlan_ip:
                 ipaddr: 1.1.1.45
                 ipmask: 255.0.0.0
       commit: True

    - name: GET the hostname of MM
      aos_api_config:
        method: GET
        api_object: hostname
        config_path: /mm/mynode

    - name: GET only profile names of SSIDs configured at /md/Boston
      aos_api_config:
        method: GET
        api_object: ssid_prof
        config_path: /md/Boston
        filter: [ {'OBJECT': { '$eq': [ 'ssid_prof.profile-name' ] } } ]




"""
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.aos_http import AosApi

def main():
    '''
    Module args and module logic
    '''
    module = AnsibleModule(
        argument_spec=dict(
            method=dict(required=True, type='str', choices=['GET', 'POST']),
            config_path=dict(required=True, type='str', default=None),
            data=dict(required=False, type='list', elements='dict', default=list()),
            commit=dict(required=False, type='bool', default=False),
            api_object=dict(required=False, type='str', default=None),
            #GET response modifiers
            filter=dict(required=False, type='list', default=list()),
            sort=dict(required=False, type='str', default=None),
            count=dict(required=False, type='str', default=None),
            limit=dict(required=False, type='int', default=None),
            offset=dict(required=False, type='int', default=None),
            total=dict(required=False, type='int', default=None),
            type=dict(required=False, type='str', default=None),

        ))

    config_path = module.params.get('config_path')
    method = module.params.get('method')
    api = AosApi(module)

    if method == "GET":
        api_object = module.params.get('api_object')
        query_url = '/configuration/object/' + api_object
        query_params = api.get_query_params()
        request = api.get_url(query_url, params=query_params)
        result = api.get(url=request)
        resp = result['resp']
        code = result['code']

        try:
            if "_data" in resp.keys():
                module.exit_json(changed=False, response=resp, response_code=code)
            elif resp is not None:
                module.exit_json(changed=False, response=resp, response_code=code)
            else:
                module.fail_json(change=False, response=resp, response_code=code,
                                 msg=str("Check if valid parameters are provided"
                                         " in the playbook."))
        except AttributeError:
            module.exit_json(changed=False, response=resp, response_code=code)

    if method == "POST":
        failed = False
        commit_status = False
        commit = module.params.get('commit')
        config_url = api.get_url('/configuration/object', params={'config_path': config_path})
        data = api.format_data()
        result, changed = api.post(url=config_url, data=data)
        resp = result['resp']
        code = result['code']
        if code == 200:
            if resp and resp != "":
                res, pending, status_str = api.validate_response(resp)
                failed = True if not res else False

                # write_memory if commit is true
                if commit and pending != 0:
                    commit_status = api.write_mem()
        else:
            failed = True

        if failed:
            module.fail_json(changed=changed, response=resp,
                             response_code=code, msg=status_str)
        else:
            module.exit_json(changed=changed, response=resp,
                             response_code=code, msg=status_str)
    else:
        module.fail_json(changed=False, msg="Invalid method type."
                         " Only GET and POST methods are supported on"
                         " Aruba Mobility Master and Controllers")

if __name__ == '__main__':
    main()
